components/core/src/clp/clg/clg.cpp

```diff
@@ -592,46 +587,24 @@ int main(int argc, char const* argv[]) {
             size_t num_bytes_read;
             file_reader.read(buf, max_map_schema_length, num_bytes_read);
             if (num_bytes_read < max_map_schema_length) {
-                auto forward_lexer_map_it = forward_lexer_map.find(buf);
-                auto reverse_lexer_map_it = reverse_lexer_map.find(buf);
+                auto lexer_map_it = lexer_map.find(buf);
                 // if there is a chance there might be a difference make a new lexer as it's pretty
                 // fast to create
-                if (forward_lexer_map_it == forward_lexer_map.end()) {
-                    // Create forward lexer
-                    auto insert_result
-                            = forward_lexer_map.emplace(buf, log_surgeon::lexers::ByteLexer());
-                    forward_lexer_ptr = &insert_result.first->second;
-                    load_lexer_from_file(schema_file_path, false, *forward_lexer_ptr);
-
-                    // Create reverse lexer
-                    insert_result
-                            = reverse_lexer_map.emplace(buf, log_surgeon::lexers::ByteLexer());
-                    reverse_lexer_ptr = &insert_result.first->second;
-                    load_lexer_from_file(schema_file_path, true, *reverse_lexer_ptr);
+                if (lexer_map_it == lexer_map.end()) {
+                    auto insert_result = lexer_map.emplace(buf, log_surgeon::lexers::ByteLexer());
+                    lexer_ptr = &insert_result.first->second;
+                    load_lexer_from_file(schema_file_path, *lexer_ptr);
                 } else {
-                    // load the lexers if they already exist
-                    forward_lexer_ptr = &forward_lexer_map_it->second;
-                    reverse_lexer_ptr = &reverse_lexer_map_it->second;
+                    lexer_ptr = &lexer_map_it->second;
                 }
             } else {
-                // Create forward lexer
-                forward_lexer_ptr = &one_time_use_forward_lexer;
-                load_lexer_from_file(schema_file_path, false, one_time_use_forward_lexer);
-
-                // Create reverse lexer
-                reverse_lexer_ptr = &one_time_use_reverse_lexer;
-                load_lexer_from_file(schema_file_path, false, one_time_use_reverse_lexer);
+                lexer_ptr = &one_time_use_lexer;
+                load_lexer_from_file(schema_file_path, one_time_use_lexer);
             }
         }
 
         // Perform search
-        if (!search(search_strings,
-                    command_line_args,
-                    archive_reader,
-                    *forward_lexer_ptr,
-                    *reverse_lexer_ptr,
-                    use_heuristic))
-        {
+        if (!search(search_strings, command_line_args, archive_reader, *lexer_ptr, use_heuristic)) {
```

Comment:
If I'm not mistaken, `*lexer_ptr` could dereference a garbage pointer if no schema file exists. This bug wasn't introduced by this PR, so we could fix it in a new PR... but we should definitely fix it.

Reply:
The same applies in `clo.cpp`.

Reply:
@coderabbitai Please create an issue to investigate and refactor the construction of lexers and how they are passed through cl{g,o,p}'s API with `use_heuristic`.